
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Entry Point &#8212; ALLCools</title>
    
  <link rel="stylesheet" href="../_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="allcools allc" href="allcools_allc.html" />
    <link rel="prev" title="Executable Documentation" href="../start/interactive.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  
  <h1 class="site-logo" id="site-title">ALLCools</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../intro.html">
   ALLCools: ALL Cytosine tools
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../start/installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../start/input_files.html">
   Input Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../start/analysis_steps.html">
   Analysis Steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../start/interactive.html">
   Executable Documentation
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Entry Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     allc
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_standard.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     standard
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_profile.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     profile
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_tbi.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_merge.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     merge
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_extract.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     extract
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_bw.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     bw
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_region.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     region
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_mcds.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcds
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="allcools_mcad.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcad
    </span>
   </code>
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  CELL LEVEL ANALYSIS
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../cell_level/basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../cell_level/step_by_step/intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../cell_level/step_by_step/100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/06-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../cell_level/step_by_step/5kb/intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/03-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../cell_level/dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../cell_level/integration/intro_integration.html">
   Data Integration
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/PreprocessRNA.html">
     RNA Data Preprocessing Before Integration
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/HarmonyIntegration.html">
     Integrating mC and RNA using Harmony
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../cell_level/doublets/intro_doublets.html">
   Doublets Identification
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  CLUSTER LEVEL ANALYSIS
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../cluster_level/intro.html">
   Cluster Level Analysis
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  VISUALIZATION
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../visualization/intro.html">
   Visualization
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  DISCUSSION
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../discuss/intro.html">
   Discussion
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../discuss/dmr_benchmark.html">
   DMR Caller Benchmark
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  API
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../api/intro.html">
   API
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../project_info/citation.html">
   Citation And Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
   <i class="fas fa-external-link-alt">
   </i>
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/command_line/allcools.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/lhqing/ALLCools"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2Fcommand_line/allcools.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/lhqing/ALLCools/edit/master/docs/allcools/command_line/allcools.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#usage">
   Usage
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="entry-point">
<h1>Entry Point<a class="headerlink" href="#entry-point" title="Permalink to this headline">¶</a></h1>
<p>All the command line tools is under the <code class="docutils literal notranslate"><span class="pre">allcools</span></code> command. The following chart illustrates their relationships.</p>
<div class="figure align-default" id="allcools-fig">
<a class="reference internal image-reference" href="../_images/ALLCools.png"><img alt="../_images/ALLCools.png" src="../_images/ALLCools.png" style="height: 600px;" /></a>
<p class="caption"><span class="caption-number">Fig. 7 </span><span class="caption-text">ALLCools command line tools.</span><a class="headerlink" href="#allcools-fig" title="Permalink to this image">¶</a></p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ allcools -h
usage: allcools <span class="o">[</span>-h<span class="o">]</span>  ...

The ALLCools <span class="nb">command</span> line toolkit contains multiple functions to manipulate the ALLC format, 
a core file format that stores single base level methylation information.
Throughout this toolkit, we use bgzip/tabix to compress and index the ALLC file to allow 
flexible data query from the ALLC file.

Current Tool List <span class="k">in</span> ALLCools:

<span class="o">[</span>Generate ALLC<span class="o">]</span>
bam-to-allc          - Generate <span class="m">1</span> ALLC file from <span class="m">1</span> position sorted BAM file via 
                       samtools mpileup.

<span class="o">[</span>Manipulate ALLC<span class="o">]</span>
standardize-allc     - Validate <span class="m">1</span> ALLC file format, standardize the chromosome names, 
                       compression format <span class="o">(</span>bgzip<span class="o">)</span> and index <span class="o">(</span>tabix<span class="o">)</span>.
tabix-allc           - A simple wrapper of tabix <span class="nb">command</span> to index <span class="m">1</span> ALLC file.
profile-allc         - Generate some summary statistics of <span class="m">1</span> ALLC
merge-allc           - Merge N ALLC files into <span class="m">1</span> ALLC file
extract-allc         - Extract information <span class="o">(</span>strand, context<span class="o">)</span> from <span class="m">1</span> ALLC file

<span class="o">[</span>Get Region Level<span class="o">]</span>
allc-to-bigwig       - Generate coverage <span class="o">(</span>cov<span class="o">)</span> and ratio <span class="o">(</span>mc/cov<span class="o">)</span> bigwig track files 
                       from <span class="m">1</span> ALLC file
allc-to-region-count - Count region level mc, cov by genome bins or provided BED files.
generate-mcds        - Generate methylation dataset <span class="o">(</span>MCDS<span class="o">)</span> <span class="k">for</span> a group of ALLC file and 
                       different region sets. This is a convenient wrapper <span class="k">function</span> <span class="k">for</span> 
                       a bunch of allc-to-region-count and xarray integration codes. 
                       MCDS is inherit from xarray.DataSet
generate-mcad        - Generate mCG hypo-methylation score AnnData dataset <span class="o">(</span>MCAD<span class="o">)</span> <span class="k">for</span> 
                       a group of ALLC file and one region set.

optional arguments:
  -h, --help            show this <span class="nb">help</span> message and <span class="nb">exit</span>

functions:
  
    allc-motif-scan <span class="o">(</span>motif<span class="o">)</span>
                        Scan a list of ALLC files using a C-Motif
                        database.C-Motif Database, can be generated via
                        <span class="s1">&#39;allcools generate-cmotif-database&#39;</span> Save the
                        integrated multi-dimensional array into netCDF4 format
                        using xarray.
    allc-to-bigwig <span class="o">(</span>bw, 2bw<span class="o">)</span>
                        Generate bigwig file<span class="o">(</span>s<span class="o">)</span> from <span class="m">1</span> ALLC file.
    allc-to-region-count <span class="o">(</span>region, 2region<span class="o">)</span>
                        Calculate mC and cov at regional level. Region can be
                        provided <span class="k">in</span> <span class="m">2</span> forms: <span class="m">1</span>. BED file, provided by
                        region_bed_paths, containing arbitrary regions and use
                        bedtools map to calculate<span class="p">;</span> <span class="m">2</span>. Fix-size non-overlap
                        genome bins, provided by bin_sizes, Form <span class="m">2</span> is much
                        faster to calculate than form <span class="m">1</span>. The output file is <span class="k">in</span>
                        <span class="m">6</span>-column bed-like format: chrom start end region_uid
                        mc cov
    ame                 Motif enrichment analysis with AME from MEME Suite.
                        See AME doc <span class="k">for</span> more information http://meme-
                        suite.org/doc/ame.html
    bam-to-allc <span class="o">(</span>allc, 2allc<span class="o">)</span>
                        Take <span class="m">1</span> position sorted BAM file, generate <span class="m">1</span> ALLC file.
    extract-allc <span class="o">(</span>extract<span class="o">)</span>
                        Extract information <span class="o">(</span>strand, context<span class="o">)</span> from <span class="m">1</span> ALLC
                        file. Able to save to several different format.
    generate-cmotif-database <span class="o">(</span>cmotif-db<span class="o">)</span>
                        Generate lookup table <span class="k">for</span> motifs all the cytosines
                        belongs to. BED files are used to limit cytosine scan
                        <span class="k">in</span> certain regions. Scanning motif over whole genome
                        is very noisy, better scan it <span class="k">in</span> some functional part
                        of genome. The result files will be <span class="k">in</span> the output
    generate-mcad <span class="o">(</span>mcad<span class="o">)</span>
                        Generate MCAD from ALLC files and one region set.
    generate-mcds <span class="o">(</span>mcds<span class="o">)</span>
                        Generate MCDS from ALLC files and region sets.
    merge-allc <span class="o">(</span>merge<span class="o">)</span>  Merge N ALLC files into <span class="m">1</span> ALLC file
    profile-allc <span class="o">(</span>profile<span class="o">)</span>
                        Generate some summary statistics of <span class="m">1</span> ALLC.
    standardize-allc <span class="o">(</span>standard<span class="o">)</span>
                        Standardize <span class="m">1</span> ALLC file by checking: <span class="m">1</span>. No header <span class="k">in</span>
                        the ALLC file<span class="p">;</span> <span class="m">2</span>. Chromosome names <span class="k">in</span> ALLC must be
                        exactly same as those <span class="k">in</span> the chrom_size_path file<span class="p">;</span> <span class="m">3</span>.
                        Output file will be bgzipped with .tbi index<span class="p">;</span> <span class="m">4</span>.
                        Remove additional chromosome
                        <span class="o">(</span><span class="nv">remove_additional_chrom</span><span class="o">=</span>True<span class="o">)</span> or raise KeyError <span class="k">if</span>
                        unknown chromosome found <span class="o">(</span>default<span class="o">)</span>
    tabix-allc <span class="o">(</span>tbi<span class="o">)</span>    a simple wrapper of tabix <span class="nb">command</span> to index <span class="m">1</span> ALLC file

Author: Hanqing Liu

See ALLCools documentation here: https://lhqing.github.io/ALLCools/intro.html
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./command_line"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="../start/interactive.html" title="previous page">Executable Documentation</a>
    <a class='right-next' id="next-link" href="allcools_allc.html" title="next page"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">allc</span></code></a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Hanqing Liu<br/>
        
            &copy; Copyright 2019 - 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>