ALLCools: ALL methyl-Cytosine tools
============================

ALLCools is a python package for single-cell DNA methylome data analysis. 
It includes functions to prepare input files, perform cellular analysis and genomic analysis.

## Documentation Organization

1. To install and get an overview, please read the **GET STARTED** part;
2. The analysis steps in ALLCools are organized in two parts:
   1. The **CELLULAR ANALYSIS** part go through the clustering, cluster DMG, and integration analysis;
   2. The **GENOMIC ANALYSIS** part go through the DMR calling, genome annotation, DNA motif analysis, correlation analysis, and enhancer prediction.
3. To generate analysis files from your own data, please read the **COMMAND LINE TOOLS** part;
4. To customize visualization, please read the **VISUALIZATION** part;
5. The **DISCUSSION** part provides more details on the analysis steps;
6. The **API** part provides reference to each function.

```{figure} ./img/doc_organize.png
---
height: 350px
name: doc-organize-fig
---
ALLCools documentation organization.
```

## Authors

- Hanqing Liu, developer, initial conception
- Jingtian Zhou, developer, 5kb clustering algorithms

## Support

```{figure} ./img/open_issue.png
---
height: 150px
name: open-issue-fig
figclass: margin
---
Click on this to create a page specific issue.
```

- The source code is on [github](https://github.com/lhqing/ALLCools);
- For releases and changelog, see [github releases page](https://github.com/lhqing/ALLCools/releases);
- For bugs and feature requests, please use the [issue tracker](https://github.com/lhqing/ALLCools/issues).
- For page-specific issues, please use the "open issue" button on the top-right toggle.

## Citation

When using ALLCools, please cite our paper {cite:p}`Liu2021` and related methods papers. 
For more details, please read the [citation page](./project_info/citation.md).
