
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ALLCools.plot.genome_track.GtfTrack</title>
    
  <link href="../../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/graphviz.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/sphinx-codeautolink.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
    <script src="../../../../_static/jquery.js"></script>
    <script src="../../../../_static/underscore.js"></script>
    <script src="../../../../_static/doctools.js"></script>
    <script src="../../../../_static/togglebutton.js"></script>
    <script src="../../../../_static/clipboard.min.js"></script>
    <script src="../../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="shortcut icon" href="../../../../_static/allcools_favicon.png"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-4F8CRF0GK9"></script>
<script>
                    window.dataLayer = window.dataLayer || [];
                    function gtag(){ dataLayer.push(arguments); }
                    gtag('js', new Date());
                    gtag('config', 'G-4F8CRF0GK9');
                </script>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../../_static/allcools_logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../intro.html">
   ALLCools: ALL methyl-Cytosine tools
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../start/installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../start/input_files.html">
   Input Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../start/analysis_steps.html">
   Analysis Steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../start/interactive.html">
   Examples In This Document
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  CELLULAR ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cell_level/basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cell_level/step_by_step/intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/100kb/06-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../cell_level/step_by_step/5kb/intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
    <label for="toctree-checkbox-4">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/5kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/5kb/02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../cell_level/step_by_step/5kb/03-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cell_level/dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cell_level/integration/intro_integration.html">
   Data Integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/integration/multi_modalities/Neurons/01-preprocess_10x.html">
     Preprocess 10X snRNA-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/integration/multi_modalities/Neurons/02-preprocess_SMART.html">
     Preprocess SMART-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/integration/multi_modalities/Neurons/03-preprocess_snATAC.html">
     Preprocess snATAC-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/integration/multi_modalities/Neurons/04-integration.html">
     Perform Cross-Modality Integration
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cell_level/doublets/intro_doublets.html">
   Doublets Identification
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cell_level/doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GENOMIC ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cluster_level/intro.html">
   Post-clustering Genomic Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/01a.call_dmr.html">
     DMR Calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/01b.other_option_to_init_region_ds.html">
     Alternative Way to Create RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/02.annotation.html">
     Annotate RegionDS
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/03.filter_dmr.html">
     Filter DMR
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cluster_level/RegionDS/intro_motif.html">
   Motif Scan and Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/04.motif_scan.html">
     Motif Scan
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/RegionDS/05.motif_enrichment.html">
     Motif Enrichment Analysis
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cluster_level/Correlation/intro_corr.html">
   Region-Region Correlation Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/Correlation/01.mc_cluster_profile.html">
     Prepare Cluster-by-Gene mC Profile
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/Correlation/02.dmr_gene_corr.html">
     DMR-Gene Correlation
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../cluster_level/REPTILE/intro_reptile.html">
   REPTILE
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/REPTILE/01.prepare_input.html">
     Prepare REPTILE
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../cluster_level/REPTILE/02.run_reptile.html">
     Predict Enhancer with REPTILE Algorithm
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools.html">
   Entry Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     allc
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_standard.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     standard
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_profile.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     profile
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_tbi.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_merge.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     merge
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_extract.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     extract
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_bw.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     bw
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_region.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     region
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_dataset.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     generate-dataset
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_mcds.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcds
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_mcad.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcad
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../command_line/allcools_table_to_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  API
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../../api/ALLCools/index.html">
   <code class="xref py py-mod docutils literal notranslate">
    <span class="pre">
     ALLCools
    </span>
   </code>
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/abc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.abc
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/>
    <label for="toctree-checkbox-13">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/abc/abc_score/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.abc.abc_score
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/clustering/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.clustering
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/>
    <label for="toctree-checkbox-14">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/doublets/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.doublets
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/>
      <label for="toctree-checkbox-15">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/clustering/doublets/coverage_doublets/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.coverage_doublets
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/clustering/doublets/scrublet/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.doublets.scrublet
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/feature_selection/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.feature_selection
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/>
      <label for="toctree-checkbox-16">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/clustering/feature_selection/feature_enrichment/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.feature_enrichment
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/clustering/feature_selection/gene_panel_design/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.clustering.feature_selection.gene_panel_design
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/ClusterMerging/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ClusterMerging
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/ConsensusClustering/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.ConsensusClustering
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/art_of_tsne/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.art_of_tsne
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/balanced_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.balanced_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/dmg/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.dmg
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/incremental_pca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.incremental_pca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/clustering/pvclust/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.clustering.pvclust
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/count_matrix/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.count_matrix
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/>
    <label for="toctree-checkbox-17">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/atac/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.atac
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/dataset/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.dataset
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/h5ad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.h5ad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/mcad/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcad
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/snap/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.snap
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/count_matrix/zarr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.count_matrix.zarr
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/dmr/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.dmr
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/>
    <label for="toctree-checkbox-18">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/dmr/call_dmr/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dmr
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/dmr/call_dms/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.call_dms
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/dmr/parse_methylpy/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.parse_methylpy
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/dmr/rms_test/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.dmr.rms_test
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/gtf/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.gtf
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/>
    <label for="toctree-checkbox-19">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/gtf/Gtf/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.Gtf
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/gtf/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.gtf.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/integration/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.integration
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/>
    <label for="toctree-checkbox-20">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/integration/cca/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.cca
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/integration/confusion/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.confusion
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/integration/harmony/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.integration.harmony
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/mcds/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.mcds
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-21" name="toctree-checkbox-21" type="checkbox"/>
    <label for="toctree-checkbox-21">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/mcds/correlation/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.correlation
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/mcds/mcds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.mcds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/mcds/region_ds/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.region_ds
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/mcds/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.mcds.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/motif/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.motif
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-22" name="toctree-checkbox-22" type="checkbox"/>
    <label for="toctree-checkbox-22">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/motif/default_motif_set/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.default_motif_set
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/motif/motifs/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.motifs
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/motif/parse_meme/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.parse_meme
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/motif/snakemake/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.motif.snakemake
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/plot/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.plot
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-23" name="toctree-checkbox-23" type="checkbox"/>
    <label for="toctree-checkbox-23">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../../api/ALLCools/plot/genome_track/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.genome_track
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-24" name="toctree-checkbox-24" type="checkbox"/>
      <label for="toctree-checkbox-24">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.GtfTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/plot/genome_track/HiCMatrixCoolTrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track.HiCMatrixCoolTrack
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/plot/genome_track/_pygenometrack/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.plot.genome_track._pygenometrack
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/categorical_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.categorical_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/color/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.color
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/continuous_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.continuous_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/contour/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.contour
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/decomposition/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.decomposition
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/dendro/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.dendro
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/interactive_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.interactive_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/qc_plots/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.qc_plots
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/size/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.size
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/sunburst/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.sunburst
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/text_anno_scatter/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.text_anno_scatter
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/plot/utilities/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.plot.utilities
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/pseudo_cell/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.pseudo_cell
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-25" name="toctree-checkbox-25" type="checkbox"/>
    <label for="toctree-checkbox-25">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/pseudo_cell/pseudo_cell_kmeans/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_kmeans
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/pseudo_cell/pseudo_cell_knn/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.pseudo_cell.pseudo_cell_knn
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/reptile/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.reptile
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-26" name="toctree-checkbox-26" type="checkbox"/>
    <label for="toctree-checkbox-26">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/reptile/reptile/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.reptile.reptile
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/sandbox/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.sandbox
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-27" name="toctree-checkbox-27" type="checkbox"/>
    <label for="toctree-checkbox-27">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.igv
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-28" name="toctree-checkbox-28" type="checkbox"/>
      <label for="toctree-checkbox-28">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/BaseClass/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.BaseClass
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/Session/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.Session
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/Track/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.Track
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/defaults/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.defaults
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/igv/make_session/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.igv.make_session
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.motif
        </span>
       </code>
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-29" name="toctree-checkbox-29" type="checkbox"/>
      <label for="toctree-checkbox-29">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/ame/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.ame
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/cmotif/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.cmotif
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/fimo/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.fimo
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/motif_scan/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.motif_scan
          </span>
         </code>
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../../api/ALLCools/sandbox/motif/utilities/index.html">
         <code class="xref py py-mod docutils literal notranslate">
          <span class="pre">
           ALLCools.sandbox.motif.utilities
          </span>
         </code>
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/sandbox/balanced_knn/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.balanced_knn
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/sandbox/scanorama/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.scanorama
        </span>
       </code>
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/sandbox/sccaf/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.sandbox.sccaf
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../../api/ALLCools/schema/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.schema
      </span>
     </code>
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-30" name="toctree-checkbox-30" type="checkbox"/>
    <label for="toctree-checkbox-30">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../../api/ALLCools/schema/mcds_schema/index.html">
       <code class="xref py py-mod docutils literal notranslate">
        <span class="pre">
         ALLCools.schema.mcds_schema
        </span>
       </code>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/__main__/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.__main__
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_allc_to_bigwig/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_allc_to_bigwig_new/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_bigwig_new
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_allc_to_region_count/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._allc_to_region_count
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_bam_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._bam_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_doc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._doc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_extract_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._extract_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_merge_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._merge_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_open/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._open
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/_version/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools._version
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/api/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.api
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/table_to_allc/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.table_to_allc
      </span>
     </code>
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../../api/ALLCools/utilities/index.html">
     <code class="xref py py-mod docutils literal notranslate">
      <span class="pre">
       ALLCools.utilities
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../../project_info/citation.html">
   Citation And Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/lhqing/ALLCools"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2F_modules/ALLCools/plot/genome_track/GtfTrack.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            
              <div>
                
  <h1>Source code for ALLCools.plot.genome_track.GtfTrack</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This GtfTrack class replaced the original pygenometracks class</span>
<span class="sd">It supports read-in a gffutils database, rather than parse very time</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">gffutils</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">font_manager</span>
<span class="kn">from</span> <span class="nn">pygenometracks.tracks.BedTrack</span> <span class="kn">import</span> <span class="n">BedTrack</span>
<span class="kn">from</span> <span class="nn">pygenometracks.tracks.GenomeTrack</span> <span class="kn">import</span> <span class="n">GenomeTrack</span>
<span class="kn">from</span> <span class="nn">pygenometracks.utilities</span> <span class="kn">import</span> <span class="n">temp_file_from_intersect</span><span class="p">,</span> <span class="n">InputError</span>

<div class="viewcode-block" id="DEFAULT_BED_COLOR"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.DEFAULT_BED_COLOR">[docs]</a><span class="n">DEFAULT_BED_COLOR</span> <span class="o">=</span> <span class="s1">&#39;#1f78b4&#39;</span></div>
<div class="viewcode-block" id="DISPLAY_BED_VALID"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.DISPLAY_BED_VALID">[docs]</a><span class="n">DISPLAY_BED_VALID</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;collapsed&#39;</span><span class="p">,</span> <span class="s1">&#39;triangles&#39;</span><span class="p">,</span> <span class="s1">&#39;interleaved&#39;</span><span class="p">,</span> <span class="s1">&#39;stacked&#39;</span><span class="p">]</span></div>
<div class="viewcode-block" id="DISPLAY_BED_SYNONYMOUS"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.DISPLAY_BED_SYNONYMOUS">[docs]</a><span class="n">DISPLAY_BED_SYNONYMOUS</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;interlaced&#39;</span><span class="p">:</span> <span class="s1">&#39;interleaved&#39;</span><span class="p">,</span> <span class="s1">&#39;domain&#39;</span><span class="p">:</span> <span class="s1">&#39;interleaved&#39;</span><span class="p">}</span></div>
<div class="viewcode-block" id="DEFAULT_DISPLAY_BED"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.DEFAULT_DISPLAY_BED">[docs]</a><span class="n">DEFAULT_DISPLAY_BED</span> <span class="o">=</span> <span class="s1">&#39;stacked&#39;</span></div>
<div class="viewcode-block" id="AROUND_REGION"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.AROUND_REGION">[docs]</a><span class="n">AROUND_REGION</span> <span class="o">=</span> <span class="mi">100000</span></div>


<div class="viewcode-block" id="_is_sqlite3"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack._is_sqlite3">[docs]</a><span class="k">def</span> <span class="nf">_is_sqlite3</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">head</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">head</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;SQLite format&#39;</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span></div>


<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;It appears you have a gene feature&quot;</span>
                                          <span class="s2">&quot; in your GTF file. You may want to use the &quot;</span>
                                          <span class="s2">&quot;`disable_infer_genes` option to speed up database &quot;</span>
                                          <span class="s2">&quot;creation&quot;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;It appears you have a transcript &quot;</span>
                                          <span class="s2">&quot;feature in your GTF file. You may want to use the &quot;</span>
                                          <span class="s2">&quot;`disable_infer_transcripts` option to speed up &quot;</span>
                                          <span class="s2">&quot;database creation&quot;</span><span class="p">)</span>
<span class="c1"># In gffutils v0.10 they changed the error message:</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;It appears you have a gene feature&quot;</span>
                                          <span class="s2">&quot; in your GTF file. You may want to use the &quot;</span>
                                          <span class="s2">&quot;`disable_infer_genes=True` option to speed up database &quot;</span>
                                          <span class="s2">&quot;creation&quot;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;It appears you have a transcript &quot;</span>
                                          <span class="s2">&quot;feature in your GTF file. You may want to use the &quot;</span>
                                          <span class="s2">&quot;`disable_infer_transcripts=True` option to speed up &quot;</span>
                                          <span class="s2">&quot;database creation&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="ReadGtf"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.ReadGtf">[docs]</a><span class="k">class</span> <span class="nc">ReadGtf</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads a gtf file.</span>

<span class="sd">    Example:</span>
<span class="sd">    gtf = ReadGtf(&quot;file.gtf&quot;)</span>
<span class="sd">    for interval in gtf:</span>
<span class="sd">        print interval.start</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">,</span> <span class="n">prefered_name</span><span class="o">=</span><span class="s2">&quot;transcript_name&quot;</span><span class="p">,</span>
                 <span class="n">merge_transcripts</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param file_path: the path of the gtf file</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">file_type</span> <span class="o">=</span> <span class="s1">&#39;bed12&#39;</span>

        <span class="c1"># list of bed fields</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fields</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;score&#39;</span><span class="p">,</span> <span class="s1">&#39;strand&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;thick_start&#39;</span><span class="p">,</span> <span class="s1">&#39;thick_end&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;rgb&#39;</span><span class="p">,</span> <span class="s1">&#39;block_count&#39;</span><span class="p">,</span>
                       <span class="s1">&#39;block_sizes&#39;</span><span class="p">,</span> <span class="s1">&#39;block_starts&#39;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">BedInterval</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;BedInterval&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">fields</span><span class="p">)</span>
        <span class="c1"># I think the name which should be written</span>
        <span class="c1"># should be the transcript_name</span>
        <span class="c1"># But we can change it to gene_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prefered_name</span> <span class="o">=</span> <span class="n">prefered_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">merge_transcripts</span> <span class="o">=</span> <span class="n">merge_transcripts</span>

        <span class="c1"># Will process the gtf to get one item per transcript:</span>
        <span class="c1"># This will create a database:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">_is_sqlite3</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureDB</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;:memory:&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">ve</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;No lines parsed&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">ve</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">all_transcripts</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">InputError</span><span class="p">(</span><span class="s2">&quot;This is not a gtf file.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">merge_transcripts</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">all_transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
                                                                <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s2">&quot;transcript&quot;</span><span class="p">)])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">all_transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s2">&quot;transcript&quot;</span><span class="p">,</span>
                                                                <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="ReadGtf.__iter__"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.ReadGtf.__iter__">[docs]</a>    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="ReadGtf.__next__"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.ReadGtf.__next__">[docs]</a>    <span class="k">def</span> <span class="fm">__next__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return: bedInterval object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_bed_interval</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">bed</span></div>

<div class="viewcode-block" id="ReadGtf.get_bed_interval"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.ReadGtf.get_bed_interval">[docs]</a>    <span class="k">def</span> <span class="nf">get_bed_interval</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Process a transcript from the database,</span>
<span class="sd">        retrieve all the values and return</span>
<span class="sd">        a namedtuple object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">tr</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">all_transcripts</span><span class="p">)</span>
        <span class="c1"># The name would be the prefered_name if exists</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">trName</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">prefered_name</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="c1"># Else try to guess the prefered_name from exons:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">trName</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">e</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">prefered_name</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                              <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span>
                              <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">tr</span><span class="p">,</span>
                                               <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;exon&#39;</span><span class="p">,</span>
                                               <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)])</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="c1"># Else take the transcript id</span>
                <span class="n">trName</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">id</span>
        <span class="c1"># If the cds is defined in the gtf,</span>
        <span class="c1"># use it to define the thick start and end</span>
        <span class="c1"># The gtf is 1-based closed intervalls</span>
        <span class="c1"># and bed are 0-based half-open so:</span>
        <span class="c1"># I need to remove one from each start</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">cds_start</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">tr</span><span class="p">,</span>
                                              <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span>
                                              <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">cds_end</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">tr</span><span class="p">,</span>
                                            <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;CDS&#39;</span><span class="p">,</span>
                                            <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;-start&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">end</span>
        <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
            <span class="c1"># If the CDS is not defined, then it is set to the start</span>
            <span class="c1"># as proposed here:</span>
            <span class="c1"># https://genome.ucsc.edu/FAQ/FAQformat.html#format1</span>
            <span class="n">cds_start</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">cds_end</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="c1"># Get all exons starts and end to get lengths</span>
        <span class="n">exons_starts</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
                        <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">tr</span><span class="p">,</span>
                                                  <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;exon&#39;</span><span class="p">,</span>
                                                  <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)]</span>
        <span class="n">exons_ends</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span><span class="o">.</span><span class="n">end</span>
                      <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">tr</span><span class="p">,</span>
                                                <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;exon&#39;</span><span class="p">,</span>
                                                <span class="n">order_by</span><span class="o">=</span><span class="s1">&#39;start&#39;</span><span class="p">)]</span>
        <span class="n">exons_length</span> <span class="o">=</span> <span class="p">[</span><span class="n">e</span> <span class="o">-</span> <span class="n">s</span> <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">e</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">exons_starts</span><span class="p">,</span> <span class="n">exons_ends</span><span class="p">)]</span>
        <span class="n">relative_exons_starts</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="o">-</span> <span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">exons_starts</span><span class="p">]</span>
        <span class="n">line_values</span> <span class="o">=</span> <span class="p">[</span><span class="n">tr</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">tr</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">tr</span><span class="o">.</span><span class="n">end</span><span class="p">,</span> <span class="n">trName</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">tr</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                       <span class="n">cds_start</span><span class="p">,</span> <span class="n">cds_end</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons_starts</span><span class="p">),</span>
                       <span class="n">exons_length</span><span class="p">,</span> <span class="n">relative_exons_starts</span><span class="p">]</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">BedInterval</span><span class="o">.</span><span class="n">_make</span><span class="p">(</span><span class="n">line_values</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="GtfTrack"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack">[docs]</a><span class="k">class</span> <span class="nc">GtfTrack</span><span class="p">(</span><span class="n">BedTrack</span><span class="p">):</span>
<div class="viewcode-block" id="GtfTrack.SUPPORTED_ENDINGS"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.SUPPORTED_ENDINGS">[docs]</a>    <span class="n">SUPPORTED_ENDINGS</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gtf&#39;</span><span class="p">,</span> <span class="s1">&#39;gtf.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;gtf.db&#39;</span><span class="p">]</span></div>
<div class="viewcode-block" id="GtfTrack.TRACK_TYPE"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.TRACK_TYPE">[docs]</a>    <span class="n">TRACK_TYPE</span> <span class="o">=</span> <span class="s1">&#39;gtf&#39;</span></div>
<div class="viewcode-block" id="GtfTrack.OPTIONS_TXT"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.OPTIONS_TXT">[docs]</a>    <span class="n">OPTIONS_TXT</span> <span class="o">=</span> <span class="n">GenomeTrack</span><span class="o">.</span><span class="n">OPTIONS_TXT</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span></div>
<span class="s2"># By default the transcript_name is used.</span>
<span class="s2"># If you want to use the gene_name:</span>
<span class="s2"># prefered_name = gene_name</span>
<span class="s2"># By default, the gtf is transformed to transcripts</span>
<span class="s2"># If you want to use see only one structure per gene</span>
<span class="s2"># merge_transcripts = true</span>
<span class="s2"># You can change the color of coding sequences by:</span>
<span class="s2">color = darkblue</span>
<span class="s2"># height of track in cm</span>
<span class="s2">height = 5</span>
<span class="s2"># whether printing the labels</span>
<span class="s2">labels = false</span>
<span class="s2"># optional:</span>
<span class="s2"># by default the labels are not printed if you have more than 60 features.</span>
<span class="s2"># to change it, just increase the value:</span>
<span class="s2">#max_labels = 60</span>
<span class="s2"># optional: font size can be given to override the default size</span>
<span class="s2">fontsize = 10</span>
<span class="s2"># optional: line_width</span>
<span class="s2">#line_width = 0.5</span>
<span class="s2"># the display parameter defines how the gtf file is plotted.</span>
<span class="s2"># Default is &#39;stacked&#39; where regions are plotted on different lines so</span>
<span class="s2"># we can see all regions and all labels.</span>
<span class="s2"># The other options are [&#39;collapsed&#39;, &#39;interleaved&#39;, &#39;triangles&#39;]</span>
<span class="s2"># These options assume that the regions do not overlap.</span>
<span class="s2"># `collapsed`: The gtf regions are plotted one after the other in one line.</span>
<span class="s2"># `interleaved`: The gtf regions are plotted in two lines, first up, then down, then up etc.</span>
<span class="s2"># optional, default is black. To remove the border, simply set &#39;border_color&#39; to none</span>
<span class="s2"># Not used in tssarrow style</span>
<span class="s2">#border_color = black</span>
<span class="s2"># style to plot the genes when the display is not triangles</span>
<span class="s2">#style = UCSC</span>
<span class="s2">#style = flybase</span>
<span class="s2">#style = tssarrow</span>
<span class="s2"># maximum number of gene rows to be plotted. This</span>
<span class="s2"># field is useful to limit large number of close genes</span>
<span class="s2"># to be printed over many rows. When several images want</span>
<span class="s2"># to be combined this must be set to get equal size</span>
<span class="s2"># otherwise, on each image the height of each gene changes</span>
<span class="s2">#gene_rows = 10</span>
<span class="s2"># by default the ymax is the number of</span>
<span class="s2"># rows occupied by the genes in the region plotted. However,</span>
<span class="s2"># by setting this option, the global maximum is used instead.</span>
<span class="s2"># This is useful to combine images that are all consistent and</span>
<span class="s2"># have the same number of rows.</span>
<span class="s2">#global_max_row = true</span>
<span class="s2"># If you want to plot all labels inside the plotting region:</span>
<span class="s2">#all_labels_inside = true</span>
<span class="s2"># If you want to display the name of the gene which goes over the plotted</span>
<span class="s2"># region in the right margin put:</span>
<span class="s2">#labels_in_margin = true</span>
<span class="s2"># if you use UCSC style, you can set the relative distance between 2 arrows on introns</span>
<span class="s2"># default is 2</span>
<span class="s2">#arrow_interval = 2</span>
<span class="s2"># if you use tssarrow style, you can choose the length of the arrow in bp</span>
<span class="s2"># (default is 4% of the plotted region)</span>
<span class="s2">#arrow_length = 5000</span>
<span class="s2"># if you use flybase or tssarrow style, you can choose the color of non-coding intervals:</span>
<span class="s2">#color_utr = grey</span>
<span class="s2"># as well as the proportion between their height and the one of coding</span>
<span class="s2"># (by default they are the same height):</span>
<span class="s2">#height_utr = 1</span>
<span class="s2"># By default, for oriented intervals in flybase style,</span>
<span class="s2"># or bed files with less than 12 columns, the arrowhead is added</span>
<span class="s2"># outside of the interval.</span>
<span class="s2"># If you want that the tip of the arrow correspond to</span>
<span class="s2"># the extremity of the interval use:</span>
<span class="s2"># arrowhead_included = true</span>
<span class="s2"># optional. If not given is guessed from the file ending.</span>
<span class="s2">file_type = </span><span class="si">{</span><span class="n">TRACK_TYPE</span><span class="si">}</span><span class="s2"></span>
<span class="s2">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="GtfTrack.DEFAULTS_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.DEFAULTS_PROPERTIES">[docs]</a>    <span class="n">DEFAULTS_PROPERTIES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;fontsize&#39;</span><span class="p">:</span> <span class="mi">12</span><span class="p">,</span>
                           <span class="s1">&#39;orientation&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                           <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="n">DEFAULT_BED_COLOR</span><span class="p">,</span>
                           <span class="s1">&#39;border_color&#39;</span><span class="p">:</span> <span class="s1">&#39;black&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;labels&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                           <span class="s1">&#39;style&#39;</span><span class="p">:</span> <span class="s1">&#39;flybase&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;display&#39;</span><span class="p">:</span> <span class="n">DEFAULT_DISPLAY_BED</span><span class="p">,</span>
                           <span class="s1">&#39;line_width&#39;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
                           <span class="s1">&#39;max_labels&#39;</span><span class="p">:</span> <span class="mi">60</span><span class="p">,</span>
                           <span class="s1">&#39;prefered_name&#39;</span><span class="p">:</span> <span class="s1">&#39;transcript_name&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;merge_transcripts&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                           <span class="s1">&#39;global_max_row&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                           <span class="s1">&#39;gene_rows&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                           <span class="s1">&#39;arrow_interval&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                           <span class="s1">&#39;arrowhead_included&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                           <span class="s1">&#39;color_utr&#39;</span><span class="p">:</span> <span class="s1">&#39;grey&#39;</span><span class="p">,</span>
                           <span class="s1">&#39;height_utr&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                           <span class="s1">&#39;arrow_length&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                           <span class="s1">&#39;region&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># Cannot be set manually but is set by tracksClass</span>
                           <span class="s1">&#39;all_labels_inside&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                           <span class="s1">&#39;labels_in_margin&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span></div>
<div class="viewcode-block" id="GtfTrack.NECESSARY_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.NECESSARY_PROPERTIES">[docs]</a>    <span class="n">NECESSARY_PROPERTIES</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;file&#39;</span><span class="p">]</span></div>
<div class="viewcode-block" id="GtfTrack.SYNONYMOUS_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.SYNONYMOUS_PROPERTIES">[docs]</a>    <span class="n">SYNONYMOUS_PROPERTIES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;display&#39;</span><span class="p">:</span> <span class="n">DISPLAY_BED_SYNONYMOUS</span><span class="p">}</span></div>
<div class="viewcode-block" id="GtfTrack.POSSIBLE_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.POSSIBLE_PROPERTIES">[docs]</a>    <span class="n">POSSIBLE_PROPERTIES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;orientation&#39;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;inverted&#39;</span><span class="p">],</span>
                           <span class="s1">&#39;style&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;flybase&#39;</span><span class="p">,</span> <span class="s1">&#39;UCSC&#39;</span><span class="p">,</span> <span class="s1">&#39;tssarrow&#39;</span><span class="p">],</span>
                           <span class="s1">&#39;display&#39;</span><span class="p">:</span> <span class="n">DISPLAY_BED_VALID</span><span class="p">}</span></div>
<div class="viewcode-block" id="GtfTrack.BOOLEAN_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.BOOLEAN_PROPERTIES">[docs]</a>    <span class="n">BOOLEAN_PROPERTIES</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;labels&#39;</span><span class="p">,</span> <span class="s1">&#39;merge_transcripts&#39;</span><span class="p">,</span> <span class="s1">&#39;global_max_row&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;arrowhead_included&#39;</span><span class="p">,</span> <span class="s1">&#39;all_labels_inside&#39;</span><span class="p">,</span>
                          <span class="s1">&#39;labels_in_margin&#39;</span><span class="p">]</span></div>
<div class="viewcode-block" id="GtfTrack.STRING_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.STRING_PROPERTIES">[docs]</a>    <span class="n">STRING_PROPERTIES</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;prefered_name&#39;</span><span class="p">,</span> <span class="s1">&#39;file&#39;</span><span class="p">,</span> <span class="s1">&#39;file_type&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;overlay_previous&#39;</span><span class="p">,</span> <span class="s1">&#39;orientation&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;title&#39;</span><span class="p">,</span> <span class="s1">&#39;style&#39;</span><span class="p">,</span> <span class="s1">&#39;color&#39;</span><span class="p">,</span> <span class="s1">&#39;border_color&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;color_utr&#39;</span><span class="p">,</span> <span class="s1">&#39;display&#39;</span><span class="p">]</span></div>
<div class="viewcode-block" id="GtfTrack.FLOAT_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.FLOAT_PROPERTIES">[docs]</a>    <span class="n">FLOAT_PROPERTIES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;fontsize&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                        <span class="s1">&#39;line_width&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                        <span class="s1">&#39;height&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                        <span class="s1">&#39;height_utr&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]}</span></div>
<div class="viewcode-block" id="GtfTrack.INTEGER_PROPERTIES"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.INTEGER_PROPERTIES">[docs]</a>    <span class="n">INTEGER_PROPERTIES</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;gene_rows&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                          <span class="s1">&#39;max_labels&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                          <span class="s1">&#39;arrow_interval&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span>
                          <span class="s1">&#39;arrow_length&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">]}</span></div>

<div class="viewcode-block" id="GtfTrack.set_properties_defaults"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.set_properties_defaults">[docs]</a>    <span class="k">def</span> <span class="nf">set_properties_defaults</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">BedTrack</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">set_properties_defaults</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fp</span> <span class="o">=</span> <span class="n">font_manager</span><span class="o">.</span><span class="n">FontProperties</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;fontsize&#39;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">colormap</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># check if the color given is a color map</span>
        <span class="c1"># Contrary to bed it cannot be a colormap</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">process_color</span><span class="p">(</span><span class="s1">&#39;color&#39;</span><span class="p">,</span> <span class="n">colormap_possible</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                           <span class="n">bed_rgb_possible</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                           <span class="n">default_value_is_colormap</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># check if border_color and color_utr are colors</span>
        <span class="c1"># if they are part of self.properties</span>
        <span class="c1"># (for example, TADsTracks do not have color_utr)</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="p">[</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;border_color&#39;</span><span class="p">,</span> <span class="s1">&#39;color_utr&#39;</span><span class="p">]</span>
                      <span class="k">if</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">process_color</span><span class="p">(</span><span class="n">param</span><span class="p">,</span> <span class="n">bed_rgb_possible</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># to set the distance between rows</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">row_scale</span> <span class="o">=</span> <span class="mf">2.3</span></div>

<div class="viewcode-block" id="GtfTrack.get_bed_handler"><a class="viewcode-back" href="../../../../api/ALLCools/plot/genome_track/GtfTrack/index.html#ALLCools.plot.genome_track.GtfTrack.GtfTrack.get_bed_handler">[docs]</a>    <span class="k">def</span> <span class="nf">get_bed_handler</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">plot_regions</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">file_to_open</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;file&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;global_max_row&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">plot_regions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">_is_sqlite3</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;file&#39;</span><span class="p">]):</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;global_max_row not supported when gtf is provided as SQLite3 db&#39;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># I do the intersection:</span>
                    <span class="n">file_to_open</span> <span class="o">=</span> <span class="n">temp_file_from_intersect</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;file&#39;</span><span class="p">],</span>
                                                            <span class="n">plot_regions</span><span class="p">,</span> <span class="n">AROUND_REGION</span><span class="p">)</span>

        <span class="n">gtf_db</span> <span class="o">=</span> <span class="n">ReadGtf</span><span class="p">(</span><span class="n">file_to_open</span><span class="p">,</span>
                         <span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;prefered_name&#39;</span><span class="p">],</span>
                         <span class="bp">self</span><span class="o">.</span><span class="n">properties</span><span class="p">[</span><span class="s1">&#39;merge_transcripts&#39;</span><span class="p">])</span>
        <span class="n">total_length</span> <span class="o">=</span> <span class="n">gtf_db</span><span class="o">.</span><span class="n">length</span>
        <span class="k">return</span> <span class="n">gtf_db</span><span class="p">,</span> <span class="n">total_length</span></div></div>
</pre></div>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Hanqing Liu<br/>
    
        &copy; Copyright 2019 - 2022.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>