
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Analysis Steps &#8212; ALLCools</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Executable Documentation" href="interactive.html" />
    <link rel="prev" title="Input Files" href="input_files.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      <h1 class="site-logo" id="site-title">ALLCools</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../intro.html">
   ALLCools: ALL Cytosine tools
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  GET STARTED
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="input_files.html">
   Input Files
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Analysis Steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interactive.html">
   Executable Documentation
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  CELL LEVEL ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/basic/intro_basic_clustering.html">
   Basic Clustering Walk-through
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mch_mcg_100k_basic.html">
     Basic Cell Clustering Using 100Kb Bins
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html">
     Basic Cell Clustering Using mCG-5Kb Bins
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/step_by_step/intro_step_by_step_clustering.html">
   Step-by-step Clustering Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../cell_level/step_by_step/100kb/intro_100kb.html">
     Step-by-step Clustering Using 100Kb bins
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/02-BasicFeatureFiltering.html">
       Feature Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/03-HighlyVariableFeatureSelection.html">
       Calculate Highly Variable Features And Get mC Fraction AnnData
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04a-PreclusteringAndClusterEnrichedFeatures-mCH.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/04b-PreclusteringAndClusterEnrichedFeatures-mCG.html">
       Preclustering and Cluster Enriched Features
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/05-DecompositionAndEmbedding.html">
       Decomposition And Embedding
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/100kb/06-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../cell_level/step_by_step/5kb/intro_5kb.html">
     Step-by-step Clustering using mCG-5Kb Input
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
    <label for="toctree-checkbox-4">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/01-CellBasicFiltering.html">
       Cell Basic Filtering
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/02-DecompositionAndEmbedding.html">
       Decomposition Using mCG-5Kb Bins
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../cell_level/step_by_step/5kb/03-Clustering.html">
       Consensus Clustering
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/dmg/intro_dmg.html">
   Differential Methylated Genes Analysis
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/01-AddGenemCFractions.html">
     Calculate Gene mC Fractions
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/02-DifferentiallyMethylatedGenes.html">
     Cluster Differentially Methylated Genes
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/03-PlotClusterDMGs.html">
     Plot Cluster DMGs
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/dmg/04-PairwiseDMG.html">
     Differential Methylated Genes - Pairwise
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/integration/intro_integration.html">
   Data Integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/01-preprocess_10x.html">
     Preprocess 10X snRNA-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/02-preprocess_SMART.html">
     Preprocess SMART-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/03-preprocess_snATAC.html">
     Preprocess snATAC-seq
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/integration/multi_modalities/Neurons/04-integration.html">
     Perform Cross-Modality Integration
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../cell_level/doublets/intro_doublets.html">
   Doublets Identification
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../cell_level/doublets/doublet_score_chrom100k.html">
     Calculate Cell Doublet Score
    </a>
   </li>
  </ul>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  CLUSTER LEVEL ANALYSIS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cluster_level/intro.html">
   Cluster Level Analysis
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  COMMAND LINE TOOLS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools.html">
   Entry Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_allc.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     allc
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_standard.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     standard
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_profile.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     profile
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_tbi.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     tbi
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_merge.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     merge
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_extract.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     extract
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_bw.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     bw
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_region.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     region
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_mcds.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcds
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../command_line/allcools_mcad.html">
   <code class="docutils literal notranslate">
    <span class="pre">
     allcools
    </span>
    <span class="pre">
     mcad
    </span>
   </code>
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  VISUALIZATION
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../visualization/intro.html">
   Visualization
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  DISCUSSION
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../discuss/intro.html">
   Discussion
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../discuss/dmr_benchmark.html">
   DMR Caller Benchmark
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  API
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../api/intro.html">
   API
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  PROJECT INFO
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../project_info/citation.html">
   Citation And Reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://github.com/lhqing/ALLCools/releases">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/start/analysis_steps.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/lhqing/ALLCools"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/lhqing/ALLCools/issues/new?title=Issue%20on%20page%20%2Fstart/analysis_steps.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/lhqing/ALLCools/edit/master/docs/allcools/start/analysis_steps.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cell-level-analysis">
   Cell Level Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sections">
     Sections
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#two-clustering-strategies-for-different-tissue-type">
     Two Clustering Strategies for different tissue type
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-using-raw-counts-from-100kb-genomic-bins">
       Clustering using raw counts from 100Kb genomic bins
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#clustering-using-hypo-methylation-score-from-5kb-genomic-bins">
       Clustering using hypo-methylation score from 5Kb genomic bins
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cluster-level-analysis">
   Cluster Level Analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     Sections
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pseudo-bulk-files">
     Pseudo-bulk Files
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="analysis-steps">
<h1>Analysis Steps<a class="headerlink" href="#analysis-steps" title="Permalink to this headline">¶</a></h1>
<p>ALLCools separates the analysis into two parts:</p>
<ol class="simple">
<li><p>Cell-level analysis, including</p>
<ul class="simple">
<li><p>Clustering analysis</p></li>
<li><p>Differentially Methylated Gene (DMG) analysis</p></li>
<li><p>Data integration</p></li>
<li><p>Identify potential doublets</p></li>
</ul>
</li>
<li><p>Cluster-level analysis, including</p>
<ul class="simple">
<li><p>Differentially Methylated Region (DMR) analysis</p></li>
<li><p>DMR annotation</p></li>
<li><p>DMR motif analysis (upstream regulators of DMRs)</p></li>
<li><p>DMR - Gene correlation analysis (downstream targets of DMRs)</p></li>
</ul>
</li>
</ol>
<p>In general, the <strong>cell-level analysis</strong> is focused on individual cells’ overall diversity
using relatively large features (kilobase-level regions or whole gene body).
One of the key results from <strong>cell-level analysis</strong> is the identification of <strong>cell clusters</strong>.
The <strong>cluster-level analysis</strong> will use these cell cluster labels to merge the cells into pseudo-bulk methylomes.
The high-coverage pseudo-bulk methylomes allow us to analyze methylation diversity at higher genome resolution
(hundred bps or even single base).</p>
<div class="section" id="cell-level-analysis">
<h2>Cell Level Analysis<a class="headerlink" href="#cell-level-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="sections">
<h3>Sections<a class="headerlink" href="#sections" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="../cell_level/basic/intro_basic_clustering.html"><span class="doc std std-doc">Basic walk-through of the clustering analysis</span></a>.</p></li>
<li><p><a class="reference internal" href="../cell_level/step_by_step/intro_step_by_step_clustering.html"><span class="doc std std-doc">Step-by-step description of the clustering analysis</span></a>.</p></li>
<li><p><a class="reference internal" href="../cell_level/dmg/intro_dmg.html"><span class="doc std std-doc">Identification of cluster Differentially Methylated Genes</span></a>.</p></li>
<li><p><a class="reference internal" href="../cell_level/integration/intro_integration.html"><span class="doc std std-doc">Cell-level data integration</span></a>.</p></li>
<li><p><a class="reference internal" href="../cell_level/doublets/intro_doublets.html"><span class="doc std std-doc">Potential cell doublets identification</span></a>.</p></li>
</ul>
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<ol class="simple">
<li><p>Cell metadata contains quality metrics (Defined by user, used in basic cell filtering).</p></li>
<li><p>Cell-by-feature raw count matrix (MCDS files) generated from single-cell ALLC files via <a class="reference internal" href="../command_line/allcools_mcds.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">mcds</span></code></span></a> for large genomic regions (e.g., 100Kb bin) and gene body counts.</p></li>
<li><p>(AND/OR) Cell-by-feature hypo-methylation score matrix (MCAD files) generated from single-cell ALLC files via
<a class="reference internal" href="../command_line/allcools_mcad.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">mcad</span></code></span></a></p></li>
</ol>
</div>
<div class="section" id="two-clustering-strategies-for-different-tissue-type">
<h3>Two Clustering Strategies for different tissue type<a class="headerlink" href="#two-clustering-strategies-for-different-tissue-type" title="Permalink to this headline">¶</a></h3>
<div class="section" id="clustering-using-raw-counts-from-100kb-genomic-bins">
<h4>Clustering using raw counts from 100Kb genomic bins<a class="headerlink" href="#clustering-using-raw-counts-from-100kb-genomic-bins" title="Permalink to this headline">¶</a></h4>
<p>In this strategy, we start from the cell-by-100kb-bin raw count matrix (<a class="reference internal" href="input_files.html#mcds-fig"><span class="std std-ref">MCDS</span></a>) for clustering analysis.
A quick demo is <a class="reference internal" href="../cell_level/basic/mch_mcg_100k_basic.html"><span class="doc std std-doc">here</span></a>, the step-by-step descriptions can be found
<a class="reference internal" href="../cell_level/step_by_step/100kb/intro_100kb.html"><span class="doc std std-doc">here</span></a>.</p>
<p>We found <span id="id1">[<a class="reference internal" href="../project_info/citation.html#id2"><span>1</span></a>]</span> using 100Kb genomic bins provides good clustering ability and is also computationally
efficient due to the relatively small number of features (27K features in mm10, v.s. 540K features using 5Kb bins).
We also tested the effect of different feature sizes in clustering adult mouse brain single-cell methylomes,
and found the size of features do not have major impact on the clustering results  <span id="id2">[<a class="reference internal" href="../project_info/citation.html#id2"><span>1</span></a>]</span>.
However, this conclusion may not apply to all tissues, as the scale of methylation diversity might be different in
different tissue or cell types. We suggest using the 100Kb bin size as a start point, and test other bin sizes when
necessary. The <a class="reference internal" href="../command_line/allcools_mcds.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">mcds</span></code></span></a> does provide the flexibility to
generate feature matrix at any bin size or through user-defined region sets.</p>
</div>
<div class="section" id="clustering-using-hypo-methylation-score-from-5kb-genomic-bins">
<h4>Clustering using hypo-methylation score from 5Kb genomic bins<a class="headerlink" href="#clustering-using-hypo-methylation-score-from-5kb-genomic-bins" title="Permalink to this headline">¶</a></h4>
<p>We also noticed that for some non-brain tissues (pituitary, PBMC etc.), the
100Kb bin clustering strategy had a hard time identifying some known cell types. One possible explanation is that the
methylation diversity of these cell types mainly occur at small discontinuous regulatory regions (DMRs) while the large
100Kb bins can only capture DMV or large-hypo DMR level diversities.
To solve this problem, we developed another strategy by using
<a class="reference internal" href="../cell_level/step_by_step/5kb/intro_5kb.html"><span class="doc std std-doc">algorithms adapted from snATAC-seq</span></a>
on small genomic bins (5Kb by default).</p>
<p>Specifically, this strategy start from cell-by-5kb-bin hypo-methylation score matrix (<a class="reference internal" href="input_files.html#mcad-fig"><span class="std std-ref">MCAD</span></a>) for clustering
analysis. A quick demo is <a class="reference internal" href="../cell_level/basic/mcg_5kb_basic.html"><span class="doc std std-doc">here</span></a>, the step-by-step descriptions can be found
<a class="reference internal" href="../cell_level/step_by_step/5kb/intro_5kb.html"><span class="doc std std-doc">here</span></a>.</p>
</div>
</div>
</div>
<div class="section" id="cluster-level-analysis">
<h2>Cluster Level Analysis<a class="headerlink" href="#cluster-level-analysis" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id3">
<h3>Sections<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Prepare pseudo-bulk ALLC files</p></li>
<li><p>Call Differentially Methylated Region (DMR)</p></li>
<li><p>DMR annotation</p></li>
<li><p>DMR motif analysis (finding upstream regulators of DMRs)</p></li>
<li><p>DMR - Gene correlation analysis (finding downstream targets of DMRs)</p></li>
</ul>
</div>
<div class="section" id="pseudo-bulk-files">
<h3>Pseudo-bulk Files<a class="headerlink" href="#pseudo-bulk-files" title="Permalink to this headline">¶</a></h3>
<p>Starting from <strong>single-cell ALLC files</strong> and <strong>cluster labels</strong> from the cell-level analysis, there are three steps to
prepare cluster-level data files.</p>
<ol class="simple">
<li><p>We create pseudo-bulk ALLC files by merging from single-cell ALLC files via
<a class="reference internal" href="../command_line/allcools_merge.html"><span class="doc std std-doc"><code class="docutils literal notranslate"><span class="pre">allcools</span> <span class="pre">merge</span></code></span></a>.</p></li>
<li><p>We then use proper algorithms (<a class="reference internal" href="../discuss/dmr_benchmark.html"><span class="doc std std-doc">see discussion</span></a>) to identify DMRs.
The expected results of this step are a DMR region list and associated statistics.</p></li>
<li><p>Using the DMR regions and additional datasets, we can annotate DMRs with multiple kinds of information, and store the
DMR-by-feature information in anndata.AnnData format, where the <code class="docutils literal notranslate"><span class="pre">obs</span></code> dimension are the DMRs, the <code class="docutils literal notranslate"><span class="pre">var</span></code> dimension are
all kinds of features.</p></li>
</ol>
<p>The data model is illustrated in the following figure:</p>
<div class="figure align-default" id="cluster-level-analysis-fig">
<a class="reference internal image-reference" href="../_images/cluster-level-analysis.png"><img alt="../_images/cluster-level-analysis.png" src="../_images/cluster-level-analysis.png" style="height: 300px;" /></a>
<p class="caption"><span class="caption-number">Fig. 6 </span><span class="caption-text">Cluster level analysis data model.</span><a class="headerlink" href="#cluster-level-analysis-fig" title="Permalink to this image">¶</a></p>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./start"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="input_files.html" title="previous page">Input Files</a>
    <a class='right-next' id="next-link" href="interactive.html" title="next page">Executable Documentation</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Hanqing Liu<br/>
        
            &copy; Copyright 2019 - 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>